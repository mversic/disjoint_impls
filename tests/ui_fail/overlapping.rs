use disjoint_impls::disjoint_impls;

pub trait Dispatch {
    type Group;
}

pub enum GroupA {}
pub enum GroupB {}
pub enum GroupC {}

impl Dispatch for String {
    type Group = GroupA;
}
impl Dispatch for i32 {
    type Group = GroupB;
}
impl Dispatch for u32 {
    type Group = GroupC;
}

disjoint_impls! {
    pub trait Kita {
        const NAME: &'static str;
    }

    impl<T: Dispatch<Group = GroupA>> Kita for T {
        const NAME: &'static str = "Blanket A";
    }
    impl<T: Dispatch<Group = GroupB>> Kita for &T {
        const NAME: &'static str = "Blanket B";
    }
    impl<U: Dispatch<Group = GroupC>> Kita for &mut U {
        const NAME: &'static str = "Blanket C";
    }
}

fn main() {}
