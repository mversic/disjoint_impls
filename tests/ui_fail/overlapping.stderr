error[E0119]: conflicting implementations of trait `Kita` for type `&_`
  --> tests/ui_fail/overlapping.rs:29:5
   |
26 |     impl<T: Dispatch<Group = GroupA>> Kita for T {
   |     -------------------------------------------- first implementation here
...
29 |     impl<T: Dispatch<Group = GroupB>> Kita for &T {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `&_`
   |
   = note: downstream crates may implement trait `Dispatch` for type `&_`

error[E0119]: conflicting implementations of trait `Kita` for type `&mut _`
  --> tests/ui_fail/overlapping.rs:32:5
   |
26 |     impl<T: Dispatch<Group = GroupA>> Kita for T {
   |     -------------------------------------------- first implementation here
...
32 |     impl<U: Dispatch<Group = GroupC>> Kita for &mut U {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `&mut _`
   |
   = note: downstream crates may implement trait `Dispatch` for type `&mut _`
